### DISCUSSION TITLE
End to end reactivity example with state on server through API and client initiated state changes

### DISCUSSION THREAD

**Discussion by edimoldovan** [solidjs/solid#1607] [Q&A] [answered]
Hi folks,

I am looking for a bit of an architectural help with regards to how to approach an end to end user journey build with Solid. Here are the steps:
- I have an empty list of elements which I retrieve from server
- I create a new element
- Save the element to the server
- Page reload on the list renders a no longer empty list
- I click one element and see element rendered in detail
- Page reload on the detail view of one of the elements renders that detail of the element

How should I approach building something like this? I read the documentation multiple times and have been searching and trying out sores and resources for days now, still can't figure out what exactly is supposed to be used for what -- do I use a store? OK, where do I put my GET and POST requests and how and where do I save their responses to my store? Do I need a store or I don't? If I don't then what do I need?

All choices I can find in the docs and guides are simply too granular and don't provide any suggestion I can follow, just a collection of options. Using a createResource for getting data is cool, but how do I combine that with a store? And how do I handle state changes as mentioned above? Do I need the createResource? Do I need the store at all?

**Comment by ryansolid** [comment, marked as answer]
Yeah it would be helpful to see more full examples I agree. Most of them are limited to only data display or only client mutation as you basically need a backend API/database for this and that often costs money and you don't want it publically available which is why you don't see demos like that.

High level you want to be using [Solid's Router](https://github.com/solidjs/solid-router) and data functions to be loading your data with resources or a data fetching library that uses resources under the hood like Tanstack Query, or Turbo Solid. Unless you have a need for a bunch of global state that isn't tied to data from the server you probably won't need a global store, but you might want some level of caching. That's what the 3rd party libraries typically add.

I think using Stores inside Resources for more optimal updates is an area that we haven't documented well and it is still considered experimental on the API side but it is being used by stuff like Tanstack Query to my knowledge. So that is probably a good start for a batteries included option. From there I think you will know if you need global state management or not. For many simple apps it just isn't really necessary with a good data fetching layer.

**Comment by edimoldovan** [comment]
Hi Ryan,

Thank for the great reply, I think Tanstack Query is what I was looking for https://github.com/TanStack/query/blob/main/examples/solid/simple/src/index.tsx -- will go ahead and use it with the Solid Router (which I'm already using).

**Comment by edimoldovan** [comment]
Would like to briefly add that Tanstack Query is working pretty much like a charm!
